File: Go/internal/data/filters.go
package data

import (
	"time"

	"github.com/google/uuid"
)

type Metadata struct {
	LastSeen       uuid.UUID `json:"lastSeen,omitzero"`
	Next           bool      `json:"next"`
	ResponseLength int       `json:"responseLength"`
}

type Filters struct {
	ID            *uuid.UUID `json:"id,omitzero"`
	OwnerID       *uuid.UUID `json:"ownerId,omitzero"`
	UserID        *uuid.UUID `json:"userId,omitzero"`
	PostID        *uuid.UUID `json:"postId,omitzero"`
	ThreadID      *uuid.UUID `json:"threadId,omitzero"`
	ForumID       *uuid.UUID `json:"forumId,omitzero"`
	AuthorID      *uuid.UUID `json:"authorId,omitzero"`
	Name          *string    `json:"name,omitzero"`
	Title         *string    `json:"title,omitzero"`
	Username      *string    `json:"username,omitzero"`
	Email         *string    `json:"email,omitzero"`
	CreatedAtFrom *time.Time `json:"createdAtFrom,omitzero"`
	CreatedAtTo   *time.Time `json:"createdAtTo,omitzero"`
	UpdatedAtFrom *time.Time `json:"updatedAtFrom,omitzero"`
	UpdatedAtTo   *time.Time `json:"updatedAtTo,omitzero"`
	DeletedAtFrom *time.Time `json:"deletedAtFrom,omitzero"`
	DeletedAtTo   *time.Time `json:"deletedAtTo,omitzero"`
	Deleted       *bool      `json:"deleted,omitzero"`
	IsLocked      *bool      `json:"isLocked,omitzero"`

	OrderBy         []string  `json:"order_by,omitzero"`
	OrderBySafeList []string  `json:"order_by_safe_list,omitzero"`
	LastSeen        uuid.UUID `json:"lastSeen,omitzero"`
	PageSize        int       `json:"page_size,omitzero"`
}

File: Python/src/python/db/filters.py
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional


@dataclass
class Filter:
    id: Optional[uuid.UUID] = None
    owner_id: Optional[uuid.UUID] = None
    user_id: Optional[uuid.UUID] = None
    post_id: Optional[uuid.UUID] = None
    thread_id: Optional[uuid.UUID] = None
    forum_id: Optional[uuid.UUID] = None
    author_id: Optional[uuid.UUID] = None
    name: Optional[str] = None
    title: Optional[str] = None
    username: Optional[str] = None
    email: Optional[str] = None
    created_at_from: Optional[datetime] = None
    created_at_to: Optional[datetime] = None
    updated_at_from: Optional[datetime] = None
    updated_at_to: Optional[datetime] = None
    deleted_at_from: Optional[datetime] = None
    deleted_at_to: Optional[datetime] = None
    deleted: Optional[bool] = None
    is_locked: Optional[bool] = None

    order_by: list[str] = field(default_factory=list)
    order_by_safelist: list[str] = field(default_factory=list)
    last_seen: uuid.UUID = uuid.UUID("00000000-0000-0000-0000-000000000000")
    page_size: int = 1

    def create_order_by_clause(self) -> str:
        if len(self.order_by) < 1:
            return "ORDER BY id"

        clauses: list[str] = []
        for clause in self.order_by:
            if clause.startswith("-"):
                clauses.append(clause.strip("-") + " DESC")
            else:
                clauses.append(clause + " ASC")

        clauses.append("id ASC")

        return "ORDER BY ".join(clauses)


@dataclass
class Metadata:
    last_seen: uuid.UUID = uuid.UUID("00000000-0000-0000-0000-000000000000")
    next: bool = False
    response_length: int = 0
